lib/full-site-editing/class-wp-rest-templates-controller.php

```diff
@@ -0,0 +1,545 @@
+		} else {
+			$result = wp_insert_post( wp_slash( (array) $changes ), true );
+		}
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+
+		$template      = gutenberg_get_block_template( $request['id'], $this->post_type );
+		$fields_update = $this->update_additional_fields_for_object( $template, $request );
+		if ( is_wp_error( $fields_update ) ) {
+			return $fields_update;
+		}
+
+		return $this->prepare_item_for_response(
+			gutenberg_get_block_template( $request['id'], $this->post_type ),
+			$request
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to create a template.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has access to create items, WP_Error object otherwise.
+	 */
+	public function create_item_permissions_check( $request ) {
+		return $this->permissions_check( $request );
+	}
+
+	/**
+	 * Creates a single template.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function create_item( $request ) {
+		$changes            = $this->prepare_item_for_database( $request );
+		$changes->post_name = $request['slug'];
+		$result             = wp_insert_post( wp_slash( (array) $changes ), true );
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+		$id = $changes->theme . '|' . $changes->slug;
+
+		$template      = gutenberg_get_block_template( $id, $this->post_type );
+		$fields_update = $this->update_additional_fields_for_object( $template, $request );
+		if ( is_wp_error( $fields_update ) ) {
+			return $fields_update;
+		}
+
+		return $this->prepare_item_for_response(
+			gutenberg_get_block_template( $id, $this->post_type ),
+			$request
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to delete a single template.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has dedlete access for the item, WP_Error object otherwise.
+	 */
+	public function delete_item_permissions_check( $request ) {
+		return $this->permissions_check( $request );
+	}
+
+	/**
+	 * Deletes a single template.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function delete_item( $request ) {
+		$template = gutenberg_get_block_template( $request['id'], $this->post_type );
+		if ( ! $template ) {
+			return new WP_Error( 'rest_template_not_found', __( 'No templates exists with that id.', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		$id    = $template->wp_id;
+		$force = (bool) $request['force'];
+
+		// If we're forcing, then delete permanently.
+		if ( $force ) {
+			$previous = $this->prepare_item_for_response( $template, $request );
+			wp_delete_post( $id, true );
```

Comment:
There's a chance that `gutenberg_get_block_template` returned a template from a theme file rather than the DB, in which case the delete here will (silently) fail. Do we need to take that into account in any way? I.e. return some kind of error msg?